File: shared/types/recognition.ts
/**
 * Shared TypeScript definitions for recognition results and vocabulary.
 *
 * These types define the contract between the recognition service and
 * clients for sign recognition outputs.
 */

/**
 * Single recognized sign result from the classifier.
 */
export interface RecognitionResult {
  /** The recognized Vietnamese sign (e.g., 'xin_chào', 'cảm_ơn') */
  sign: string;
  /** Confidence score from the model (0.0 to 1.0) */
  confidence: number;
}

/**
 * Phrase completion event - sent when a phrase boundary is detected.
 */
export interface PhraseComplete {
  /** The complete recognized text (concatenated signs) */
  text: string;
  /** Individual sign results that make up the phrase */
  signs: RecognitionResult[];
  /** Optional base64-encoded audio for TTS output */
  audio?: string;
}

/**
 * Sign vocabulary entry for model output mapping.
 */
export interface SignVocabulary {
  /** Unique identifier for the sign */
  id: string;
  /** Vietnamese text representation (snake_case) */
  text: string;
  /** Optional category for grouping (e.g., 'greetings', 'pronouns') */
  category?: string;
}

/**
 * Vocabulary collection format for loading/saving.
 */
export interface VocabularyCollection {
  /** Version string for vocabulary schema */
  version: string;
  /** Array of sign vocabulary entries */
  signs: SignVocabulary[];
}

/**
 * Recognition service health status.
 */
export interface HealthStatus {
  status: 'healthy' | 'unhealthy';
  modelLoaded: boolean;
  vocabularySize: number;
  uptime: number;
}

File: recognition-service/app/schemas/recognition.py
"""
Recognition-related Pydantic schemas.
"""
from pydantic import BaseModel, field_validator
from typing import List, Optional, Dict, Any


class RecognitionResult(BaseModel):
    """Result of a single sign recognition."""
    sign: str
    confidence: float

    @field_validator('confidence')
    @classmethod
    def check_confidence_range(cls, v):
        if not 0.0 <= v <= 1.0:
            raise ValueError('Confidence must be between 0.0 and 1.0')
        return v


class PhraseComplete(BaseModel):
    """Event sent when a phrase is complete and ready for TTS."""
    text: str
    signs: List[RecognitionResult]
    audio: Optional[str] = None  # Base64 encoded WAV audio

    def has_audio(self) -> bool:
        """Check if this event includes audio data."""
        return self.audio is not None and len(self.audio) > 0


class TTSErrorResponse(BaseModel):
    """Error response when TTS fails."""
    provider: str
    message: str
    fallbackAttempted: Optional[bool] = False
